# comparative-genomics-project

Main directory structure:
```text
.
├── code/      # -- analysis code/notebooks (after pipeline run) => R/python
├── data/      # -- data files
├── dev/       # -- development scripts, tools, installation scripts (reproducibility and logging)
├── docs/      # -- documentation files
├── logs/      # -- log files and version
├── readme.md  # -- project overview and instructions
├── output/    # -- output files and results
├── pipeline/  # -- pipeline scripts and workflow
└── scripts/   # -- command line scripts
```

> [!IMPORTANT]  
> To mark a change you've added or done to the project, please commit with a clear message on what was changed and run  
> ``` ./dev/version_tracker.sh ```, follow the instructions (choose an option - default z unless you did a big change, write each step you've done in a seperate line and the press `ENTER` to get out of the prompt => version will be updated automatically)


## Data

The main data files to that will be found in `data/` can be extracted this way:

1. **Whole Glycine max genome (genes FASTA)**
```bash
curl http://ftp.ensemblgenomes.org/pub/release-41/plants/fasta/glycine_max/dna_index/Glycine_max.Glycine_max_v2.0.dna.toplevel.fa.gz -o data/genome.fa.gz
gunzip data/genome.fa.gz 
```
But I think what we need is the peptide sequences:
```bash
curl http://ftp.ensemblgenomes.org/pub/release-41/plants/fasta/glycine_max/pep/Glycine_max.Glycine_max_v2.0.pep.all.fa.gz -o data/peptides.fa.gz
gunzip data/peptides.fa.gz
```

`data/protein_info.csv` is generated by the `scripts/0_extract_data.sh` script, it contains metadata about each protein in the FASTA file

## Dependencies

_Check out installation in [scripts/SETUP.sh](scripts/SETUP.sh)_

```
ncbi-blast+
```