
## Data

The main data files to that will be found in `data/` can be extracted this way:

1. **Whole Glycine max genome (genes FASTA)**
<!-- ```bash
curl http://ftp.ensemblgenomes.org/pub/release-41/plants/fasta/glycine_max/dna_index/Glycine_max.Glycine_max_v2.0.dna.toplevel.fa.gz -o data/genome.fa.gz
gunzip data/genome.fa.gz 
``` -->
But I think what we need is the peptide sequences:
```bash
curl http://ftp.ensemblgenomes.org/pub/release-41/plants/fasta/glycine_max/pep/Glycine_max.Glycine_max_v2.0.pep.all.fa.gz -o data/peptides.fa.gz
gunzip data/peptides.fa.gz
```

`data/protein_info.csv` is generated by the `scripts/0_extract_data.sh` script, it contains metadata about each protein in the FASTA file

2. **Coding sequences (CDS FASTA)**
For Ks calculation we need the CDS sequences. They can be downloaded from Ensembl Plants:
```bash
curl http://ftp.ensemblgenomes.org/pub/release-41/plants/fasta/glycine_max/cdna/Glycine_max.Glycine_max_v2.0.cdna.all.fa.gz -o data/cds.fa.gz
gunzip data/cds.fa.gz
```
